File: Anul II/Medii de proiectare si programare/Laborator/lab03/C#/Ubb.BikeContest.Repository/src/RaceDbRepository.cs
using System.Collections;
using System.Data;
using log4net;
using Ubb.BikeContest.Model;

namespace Ubb.BikeContest.Repository;

public class RaceDbRepository : IRaceRepository
{
    private static readonly ILog Log = LogManager.GetLogger("RaceDbRepository");
    private readonly IDictionary<string, string?> _props;

    public RaceDbRepository(IDictionary<string, string?> props)
    {
        Log.Info("Creating RaceDbRepository ");
        _props = props;
    }

    public Race FindById(long id)
    {
        Log.InfoFormat("Entering Get with value {0}", id);
        var connection = DbUtils.GetConnection(_props);

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "SELECT id,name,engineCc FROM races WHERE id=@id";
            var paramId = command.CreateParameter();
            paramId.ParameterName = "@id";
            paramId.Value = id;
            command.Parameters.Add(paramId);

            using (var dataReader = command.ExecuteReader())
            {
                if (dataReader.Read())
                {
                    Race race = Extract(dataReader);
                    Log.InfoFormat("Exiting Get with value {0}", race);
                    return race;
                }
            }
        }
        Log.InfoFormat("Exiting Get with value {0}", null);
        return null;
    }

    public IEnumerable FindAll()
    {
        Log.InfoFormat("Entering Read");
        var connection = DbUtils.GetConnection(_props);
        IList<Race> races = new List<Race>();

        using (var command = connection.CreateCommand())
        {
            command.CommandText = "SELECT id,name,engineCc FROM races";
            using (var dataReader = command.ExecuteReader())
            {
                while (dataReader.Read())
                {
                    Race race = Extract(dataReader);
                    races.Add(race);
                }
            }
        }
        Log.InfoFormat("Exiting Read with value {0}", races);
        return races;
    }

    public void Save(Race newEntity)
    {
        Log.InfoFormat("Entering Add with value {0}", newEntity);
        var connection = DbUtils.GetConnection(_props);

        using var command = connection.CreateCommand();
        command.CommandText = 
            "INSERT INTO races(name, engineCc) " +
            "VALUES(@name, @engineCc);";
        var name = command.CreateParameter();
        name.ParameterName = "@name";
        name.Value = newEntity.Name;
        command.Parameters.Add(name);

        var engineCapacity = command.CreateParameter();
        engineCapacity.ParameterName = "@engineCc";
        engineCapacity.Value = newEntity.EngineCapacity;
        command.Parameters.Add(engineCapacity);

        var result = command.ExecuteNonQuery();
        Log.InfoFormat("Added {0} entities", result);
    }

    public void Delete(long id)
    {
        Log.InfoFormat("Entering Delete with value {0}", id);
        IDbConnection connection = DbUtils.GetConnection(_props);
        using var command = connection.CreateCommand();
        command.CommandText = "DELETE FROM races WHERE id=@id";
        var paramId = command.CreateParameter();
        paramId.ParameterName = "@id";
        paramId.Value = id;
        command.Parameters.Add(paramId);
        var result = command.ExecuteNonQuery();
        Log.InfoFormat("Deleted {0} entities", result);
    }

    public void Update(Race updatedEntity)
    {
        Log.InfoFormat("Entering Update with value {0}", updatedEntity);
        IDbConnection connection = DbUtils.GetConnection(_props);

        using var command = connection.CreateCommand();
        command.CommandText = 
            "UPDATE races SET name=@name, engineCc=@engineCc WHERE id=@id";
        var name = command.CreateParameter();
        name.ParameterName = "@name";
        name.Value = updatedEntity.Name;
        command.Parameters.Add(name);

        var engineCapacity = command.CreateParameter();
        engineCapacity.ParameterName = "@engineCc";
        engineCapacity.Value = updatedEntity.EngineCapacity;
        command.Parameters.Add(engineCapacity);

        var id = command.CreateParameter();
        id.ParameterName = "@id";
        id.Value = updatedEntity.Id;
        command.Parameters.Add(id);

        var result = command.ExecuteNonQuery();
        Log.InfoFormat("Updated {0} entities", result);
    }

    public Race GetRaceByName(string name)
    {
        Log.InfoFormat("Entering GetRaceByName with value {0}", name);
        IDbConnection connection = DbUtils.GetConnection(_props);

        using var command = connection.CreateCommand();
        command.CommandText = "SELECT id,name,engineCc FROM races WHERE name=@name";
        var paramName = command.CreateParameter();
        paramName.ParameterName = "@name";
        paramName.Value = name;
        command.Parameters.Add(paramName);

        using (var dataReader = command.ExecuteReader())
        {
            if (dataReader.Read())
            {
                Race race = Extract(dataReader);
                Log.InfoFormat("Exiting GetRaceByName with value {0}", race);
                return race;
            }
        }

        Log.InfoFormat("Exiting GetRaceByName with value {0}", null);
        return null;
    }

    private Race Extract(IDataReader dataReader)
    {
        var id = dataReader.GetInt64(0);
        var name = dataReader.GetString(1);
        var engineCapacity = dataReader.GetInt32(2);

        var race = new Race(name, engineCapacity)
        {
            Id = id
        };
        return race;
    }
}

File: Anul II/Medii de proiectare si programare/Laborator/lab04/src/main/java/com/ubb/repository/RaceDBRepository.java
package com.ubb.repository;

import com.ubb.model.Race;
import com.ubb.utils.JDBCUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Optional;
import java.util.Properties;

public class RaceDBRepository implements IRaceRepository {
    private final JDBCUtils jdbcUtils;
    private static final Logger logger = LogManager.getLogger();

    public RaceDBRepository(final Properties properties) {
        logger.info("Initializing RaceDBRepository with properties: {} ", properties);
        jdbcUtils = new JDBCUtils(properties);
    }

    @Override
    public Optional<Race> getRaceByName(String name) {
        logger.traceEntry("getRaceByName with task {} ", name);
        Connection connection = jdbcUtils.getConnection();
        Collection<Race> races = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM races WHERE name=?")) {
            preparedStatement.setString(1, name);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                if (resultSet.next()) {
                    Optional<Race> race = Optional.of(extract(resultSet));
                    logger.traceExit(race.get());
                    return race;
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit("null");
        return Optional.empty();
    }

    @Override
    public Collection<Race> getRacesByEngineCapacity(Integer engineCapacity) {
        logger.traceEntry("getRacesByEngineCapacity with task {} ", engineCapacity);
        Connection connection = jdbcUtils.getConnection();
        Collection<Race> races = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM races WHERE engineCc=?")) {
            preparedStatement.setInt(1, engineCapacity);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Race race = extract(resultSet);
                    races.add(race);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit(races);
        return races;
    }

    @Override
    public Optional<Race> findById(Long id) {
        logger.traceEntry();
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM races WHERE id=?")) {
            preparedStatement.setLong(1, id);
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                if (resultSet.next()) {
                    Race race = extract(resultSet);
                    logger.traceExit(race);
                    return Optional.of(race);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit("null");
        return Optional.empty();
    }

    @Override
    public Collection<Race> findAll() {
        logger.traceEntry();
        Connection connection = jdbcUtils.getConnection();
        Collection<Race> races = new ArrayList<>();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "SELECT * FROM races")) {
            try (ResultSet resultSet = preparedStatement.executeQuery()) {
                while (resultSet.next()) {
                    Race race = extract(resultSet);
                    races.add(race);
                }
            }
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
        logger.traceExit(races);
        return races;
    }

    @Override
    public void save(Race newEntity) {
        logger.traceEntry("saving task {} ", newEntity);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "INSERT INTO races (name, engineCc) VALUES (?,?)")) {
            preparedStatement.setString(1, newEntity.getName());
            preparedStatement.setInt(2, newEntity.getEngineCapacity());
            final int result = preparedStatement.executeUpdate();
            logger.traceExit("Saved {} instances ", result);
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
    }

    @Override
    public void delete(Long id) {
        logger.traceEntry("deleting task of id={} ", id);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "DELETE FROM races WHERE id=?")) {
            preparedStatement.setLong(1, id);
            final int result = preparedStatement.executeUpdate();
            logger.traceExit("Deleted {} instances", result);
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
    }

    @Override
    public void update(Race updatedEntity) {
        logger.traceEntry("updating with task {} ", updatedEntity);
        Connection connection = jdbcUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement(
                "UPDATE races SET name=?, engineCc=? where id=?")) {
            preparedStatement.setString(1, updatedEntity.getName());
            preparedStatement.setInt(2, updatedEntity.getEngineCapacity());
            preparedStatement.setLong(3, updatedEntity.getID());
            int result = preparedStatement.executeUpdate();
            logger.traceExit("Updated {} instances");
        } catch (SQLException sqlException) {
            logger.error(sqlException);
            System.err.println("DB Error : " + sqlException);
        }
    }

    private Race extract(ResultSet resultSet) throws SQLException {
        Long id = resultSet.getLong("id");
        String name = resultSet.getString("name");
        Integer engineCapacity = resultSet.getInt("engineCc");

        Race race = new Race(name, engineCapacity);
        race.setID(id);
        return race;
    }
}
